### PR TITLE
⚙️ [codex-plugin-stability-feedback] fix(codex): unify code-mode command identity [step 2/11]

### PR DESCRIPTION
**Pull request by AlexDochioiu**
## Complexity

Moderate (`⚙️`): this changes per-turn identity correlation across Codex rollout classification and lifecycle tracking, with focused state and replay edge cases contained inside the Codex plugin.

## What

- Recognize persisted code-mode `exec` calls that contain one `tools.exec_command(...)` invocation.
- Correlate that rollout call with its app-server command item through a dedicated per-turn queue, preserving one canonical tool ID through completion.
- Clear the additional correlation state with the existing lifecycle reset and add a focused regression.
- Advance the durable stability plan and tracker to Step 2/11.

## Why

An observed code-mode shell command could appear as two live tool cards because the rollout `call_*` ID and app-server `exec-*` ID were treated independently. Reload later collapsed the duplicate, but live session history was inconsistent.

## Risk and test focus

Medium risk, limited to Codex code-mode command identity. Review and CI should focus on single-command correlation, existing direct shell correlation, generated wrapper classification, completion state, and reset behavior. OpenCode, Cursor, wire contracts, and database persistence are unchanged.

## Expected result

A single code-mode shell command keeps one canonical tool card and ID from its rollout record through app-server start and completion. There is no database or transport change and no behavior change for other backend plugins.

## Verification

- `dart test test/codex_tool_lifecycle_tracker_test.dart`
- `dart test` — all 298 Codex package tests passed
- `dart analyze --fatal-infos` — no issues
- `git diff --check origin/main...HEAD`

---
## Summary by cubic
Unifies code‑mode shell command identity so one tool card persists from rollout `exec` to app-server start and completion. Hardens detection to only correlate a single real `tools.exec_command(...)` call, ignoring strings and comments.

- **Bug Fixes**
  - Parse code-mode `exec` input to find exactly one `tools.exec_command(...)` call outside strings/comments; accept invocation whitespace.
  - Correlate that rollout call with app-server `exec-*` via a per-turn queue to keep one canonical tool ID through completion.
  - Ignore invocation-like text inside strings; add tests for single, whitespace, and embedded cases.
  - Clear correlation state on lifecycle reset and update Step 2/11 plan/tracker.

<sup>Written for commit 70ae07cb13da5d43e90f20bf9d004ae9851f390c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/sesori-ai/sesori_apps_monorepo/pull/731?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: bridge/sesori_plugin_codex/lib/src/repositories/mappers/codex_rollout_tool_mapper.dart [hunk 1]
```diff
@@ -132,6 +132,20 @@ class CodexRolloutToolMapper {
...
+    required CodexRolloutResponseItemDto payload,
+  }) {
+    if (payload case CodexRolloutCustomToolCallDto(
+      :final name,
+      :final input,
+    ) when name.toLowerCase() == "exec") {
+      const marker = "tools.exec_command(";
+      final first = input.indexOf(marker);
+      return first >= 0 && input.indexOf(marker, first + marker.length) < 0;
+    }
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_codex/lib/src/repositories/mappers/codex_rollout_tool_mapper.dart]: [Sesori reply] Addressed (in commit 70ae07cb)

The classifier now ignores `tools.exec_command` text inside strings and comments and requires a balanced real invocation. A focused regression verifies that literal-only text does not enqueue correlation.

File: bridge/sesori_plugin_codex/lib/src/repositories/codex_tool_lifecycle_tracker.dart [hunk 2]
```diff
@@ -120,6 +120,16 @@ class CodexToolLifecycleTracker {
         if (pending.isEmpty) thread.pendingShellCallsByTurn.remove(turnId);
       }
     }
+    if (canonicalId == null && turnId != null) {
+      final pending = thread.pendingCodeModeShellCallsByTurn[turnId];
+      if (pending != null && pending.isNotEmpty) {
+        canonicalId = pending.removeAt(0);
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_codex/lib/src/repositories/codex_tool_lifecycle_tracker.dart:126]: [Sesori reply] Not addressed

The separate queues intentionally make code-mode correlation a fallback when no direct rollout `exec_command` candidate exists. A shared FIFO would let an outer custom wrapper claim the app-server item before its authoritative direct rollout call, regressing the existing wrapper/direct flow; the mixed distinct-command batching case is not evidenced.

File: bridge/sesori_plugin_codex/lib/src/repositories/mappers/codex_rollout_tool_mapper.dart [hunk 3]
```diff
@@ -132,6 +132,20 @@ class CodexRolloutToolMapper {
...
+  bool isSingleCodeModeCommandExecutionCall({
+    required CodexRolloutResponseItemDto payload,
+  }) {
+    if (payload case CodexRolloutCustomToolCallDto(
+      :final name,
+      :final input,
+    ) when name.toLowerCase() == "exec") {
+      const marker = "tools.exec_command(";
+      final first = input.indexOf(marker);
+      return first >= 0 && input.indexOf(marker, first + marker.length) < 0;
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_codex/lib/src/repositories/mappers/codex_rollout_tool_mapper.dart]: [Sesori reply] Addressed (in commit 70ae07cb)

Replaced raw marker counting with a quote/comment-aware scan that requires one balanced invocation. Invocation-like text inside the command argument no longer creates a second match, and the new regression covers that case.

File: bridge/sesori_plugin_codex/lib/src/repositories/mappers/codex_rollout_tool_mapper.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_codex/lib/src/repositories/mappers/codex_rollout_tool_mapper.dart]: [Sesori reply] Addressed (in commit 70ae07cb)

The invocation matcher now accepts whitespace before the opening parenthesis, with a focused correlation regression for `tools.exec_command (...)`.

File: bridge/sesori_plugin_codex/lib/src/repositories/codex_tool_lifecycle_tracker.dart [hunk 4]
```diff
@@ -120,6 +120,16 @@ class CodexToolLifecycleTracker {
         if (pending.isEmpty) thread.pendingShellCallsByTurn.remove(turnId);
       }
     }
+    if (canonicalId == null && turnId != null) {
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_codex/lib/src/repositories/codex_tool_lifecycle_tracker.dart:123]: [Sesori reply] Not addressed

This is the same mixed-order suggestion assessed in the prior thread. Code-mode candidates are deliberately fallback-only: when a direct rollout `exec_command` exists, it is the authoritative match. There is no shared app-server/rollout discriminator, and one cross-type FIFO would let an outer custom wrapper steal the app-server item from its direct call, regressing the observed wrapper/direct flow to protect an unobserved mixed-form batching case.

File: bridge/sesori_plugin_codex/lib/src/repositories/codex_tool_lifecycle_tracker.dart [hunk 5]
```diff
@@ -238,8 +248,16 @@ class CodexToolLifecycleTracker {
...
       tool.title ??= call.title;
-      if (effectiveTurnId != null && _rolloutToolMapper.isCommandExecutionCall(payload: payload)) {
-        final pending = thread.pendingShellCallsByTurn.putIfAbsent(effectiveTurnId, () => []);
+      Map<String, List<String>>? pendingByTurn;
+      if (_rolloutToolMapper.isCommandExecutionCall(payload: payload)) {
+        pendingByTurn = thread.pendingShellCallsByTurn;
+      } else if (_rolloutToolMapper.isSingleCodeModeCommandExecutionCall(
+        payload: payload,
+      )) {
+        pendingByTurn = thread.pendingCodeModeShellCallsByTurn;
```
[the remark this answers is not in the document]
Comment (AlexDochioiu) [reply bridge/sesori_plugin_codex/lib/src/repositories/codex_tool_lifecycle_tracker.dart:257]: [Sesori reply] Not addressed

A custom-tool result proves the wrapper finished, but not that its app-server notification cannot arrive later; pruning on that result would reintroduce the duplicate identity for delayed notifications, an observed lifecycle shape elsewhere in this stability pass. No ordinary-flow evidence shows dead-branch wrappers followed by another same-turn code-mode command, so adding another timing heuristic is not justified without a backend correlation key or a reproduced user impact.
